File: csharp/csvviewer/csvviewer/src/csvviewer/Interactors.cs
using System.Collections.Generic;

namespace csvviewer
{
    public class Interactors
    {
        private readonly Paging paging = new Paging();
        private readonly CommandLine commandLine = new CommandLine();
        private readonly FileProvider fileProvider = new FileProvider();

        public IEnumerable<Record> Start(string[] args) {
            var filename = commandLine.GetFilename(args);
            var pageLength = commandLine.GetPageLength(args);
            var lines = fileProvider.ReadFileContent(filename);
            var firstPage = paging.ExtractFirstPage(lines, pageLength);
            var records = Csv.CreateRecords(firstPage);
            return records;
        }

        public IEnumerable<Record> MoveNext() {
            var pageLength = commandLine.GetPageLength();
            var lines = fileProvider.GetFileContent();
            var nextPage = paging.ExtractNextPage(lines, pageLength);
            var records = Csv.CreateRecords(nextPage);
            return records;
        }
    }
}

File: cpp/csvviewer/csvviewer/Interactors.cpp
#include "Interactors.h"
#include "Csv.h"

namespace csvviewer 
{
    std::vector<Record> Interactors::Start(std::vector<std::string> args) {
        auto filename = commandline.GetFilename(args);
        auto pageLength = commandline.GetPageLength(args);
        auto lines = fileProvider.ReadFileContent(filename);
        auto firstPage = paging.ExtractFirstPage(lines, pageLength);
        auto records = Csv::CreateRecords(firstPage);
        return records;
    }

    std::vector<Record> Interactors::FirstPage() {
        auto pageLength = commandline.GetPageLength();
        auto lines = fileProvider.GetFileContent();
        auto firstPage = paging.ExtractFirstPage(lines, pageLength);
        auto records = Csv::CreateRecords(firstPage);
        return records;
    }

    std::vector<Record> Interactors::PrevPage() {
        auto pageLength = commandline.GetPageLength();
        auto lines = fileProvider.GetFileContent();
        auto prevPage = paging.ExtractPrevPage(lines, pageLength);
        auto records = Csv::CreateRecords(prevPage);
        return records;
    }

    std::vector<Record> Interactors::NextPage() {
        auto pageLength = commandline.GetPageLength();
        auto lines = fileProvider.GetFileContent();
        auto nextPage = paging.ExtractNextPage(lines, pageLength);
        auto records = Csv::CreateRecords(nextPage);
        return records;
    }

    std::vector<Record> Interactors::LastPage() {
        auto pageLength = commandline.GetPageLength();
        auto lines = fileProvider.GetFileContent();
        auto lastPage = paging.ExtractLastPage(lines, pageLength);
        auto records = Csv::CreateRecords(lastPage);
        return records;
    }
}